python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1380 @@
+            self.log.info('Finished gain estimation for detector %s' % detNum)
+        else:
+            gains = dataRef.get('brighterFatterGain')
+            if not gains:
+                raise RuntimeError('Failed to retrieved gains for detector %s' % detNum)
+            self.log.info('Retrieved stored gain for detector %s' % detNum)
+        self.log.debug('Detector %s has gains %s' % (detNum, gains))
+
+        # Loop over pairs of visits
+        # calculating the cross-correlations at the required level
+        for (v1, v2) in visitPairs:
+
+            dataRef.dataId['visit'] = v1
+            exp1 = self.isr.runDataRef(dataRef).exposure
+            dataRef.dataId['visit'] = v2
+            exp2 = self.isr.runDataRef(dataRef).exposure
+            del dataRef.dataId['visit']
+            self._checkExpLengthEqual(exp1, exp2, v1, v2)
+
+            self.log.info('Preparing images for cross-correlation calculation for detector %s' % detNum)
+            # note the shape of these returns depends on level
+            _scaledMaskedIm1, _means1 = self._prepareImage(exp1, gains, self.config.level)
+            _scaledMaskedIm2, _means2 = self._prepareImage(exp2, gains, self.config.level)
+
+            if self.debug.enabled:
+                try:
+                    frameId1 = exp1.getMetadata().get("FRAMEID")
+                    frameId2 = exp2.getMetadata().get("FRAMEID")
+                    frameId = '_diff_'.join(frameId1, frameId2)
+                except:
+                    frameId = 'Im1 diff Im2'
+
+            # Compute the cross-correlation and means
+            # at the appropriate config.level:
+            # - "DETECTOR": one key, so compare the two visits to each other
+            # - "AMP": n_amp keys, comparing each amplifier of one visit
+            #          to the same amplifier in the visit its paired with
+            for det_object in _scaledMaskedIm1.keys():
+                xcorrs[det_object].append(self._crossCorrelate(_scaledMaskedIm1[det_object],
+                                                               _scaledMaskedIm2[det_object]))
+                means[det_object].append([_means1[det_object], _means2[det_object]])
+
+                # TODO: DM-15305 improve debug functionality here.
+                # This is position 1 for the removed code.
+
+        # generate the kernel(s)
+        self.log.info('Generating kernel(s) for %s' % detNum)
+        for det_object in xcorrs.keys():  # looping over either detectors or amps
+            if self.config.level == 'DETECTOR':
+                objId = 'detector %s' % det_object
+            elif self.config.level == 'AMP':
+                objId = 'detector %s AMP %s' % (detNum, det_object)
+            kernels[det_object] = self.generateKernel(xcorrs[det_object], means[det_object], objId)
+        # writing the level might be useful, but might also mess things up xxx
+        # kernels['level'] = self.config.level
```

Comment:
Please don't leave in commented-out code (at least not without a ticket #). It is an interesting issue. It seems like a great idea to include the info, but it's weird to mix data with metadata.

Reply:
I think both of those are good suggestions, and I especially like the latter. However, I will just remove for now, and if/when we decide we want that I can make a ticket and new datasets etc, as we're not really losing any code here.